torchrl/collectors/distributed/ray_collector.py

```diff
@@ -0,0 +1,321 @@
+
+            # Retrieve and concatenate Tensordicts
+            out_td = []
+            for r in pending_tasks:
+                rollouts = ray.get(r)
+                ray.internal.free(r)  # should not be necessary, deleted automatically when ref count is down to 0
+                out_td.append(rollouts)
+            if len(rollouts.batch_size):
+                out_td = torch.stack(out_td)
+            else:
+                out_td = torch.cat(out_td)
+
+            self.collected_frames += out_td.numel()
+
+            yield out_td
+
+        self.stop_remote_collectors()
+        ray.shutdown()
+
+    def _async_iterator(self) -> Iterator[TensorDictBase]:
+
+        pending_tasks = {}
+        for w in self.remote_collectors():
+            future = w._iterator_step.remote()
+            pending_tasks[future] = w
+
+        while self.collected_frames < self.total_frames:
+
+            if not len(list(pending_tasks.keys())) == len(self.remote_collectors()):
+                raise RuntimeError("Missing pending tasks, something went wrong")
+
+            # Wait for first worker to finish
+            wait_results = ray.wait(object_refs=list(pending_tasks.keys()))
+            future = wait_results[0][0]
+            w = pending_tasks.pop(future)
+
+            # Retrieve single rollouts
+            out_td = ray.get(future)
+            ray.internal.free(future)  # should not be necessary, deleted automatically when ref count is down to 0
+            self.collected_frames += out_td.numel()
+
+            # Update agent weights
+            policy_weights_local_collector = {"policy_state_dict": self.local_policy.state_dict()}
+            policy_weights_local_collector_ref = ray.put(policy_weights_local_collector)
+            w.load_state_dict.remote(**{"state_dict": policy_weights_local_collector_ref, "strict": False})
+
+            # Schedule a new collection task
+            future = w._iterator_step.remote()
+            pending_tasks[future] = w
+
+            yield out_td
+
+        # Wait for the in-process collections tasks to finish.
+        refs = list(pending_tasks.keys())
+        ray.wait(object_refs=refs, num_returns=len(refs))
+
+        # Cancel the in-process collections tasks
+        # for ref in refs:
+        #     ray.cancel(
+        #         object_ref=ref,
+        #         force=False,
+        #     )
+
+        self.stop_remote_collectors()
+        ray.shutdown()
+
+    def set_seed(self, seed: Union[int, List[int]], static_seed: bool = False) -> List[int]:
+        """Calls set_seed(seed, static_seed) method for each remote collector and results a list of results."""
+        if isinstance(seed, int):
+            seed = [seed]
+        if len(seed) == 1:
+            seed = seed * len(self.remote_collectors())
+        futures = [collector.set_seed.remote(s, static_seed) for s, collector in zip(seed, self.remote_collectors())]
+        results = ray.get(object_refs=futures)
+        return results
```

Comment:
we should set the seed iteratively
In torchrl seeding is done via a generator that gives you the next seed to use when you need to set it multiple times
Like this you can set different seeds across workers whilst keeping the "deterministic" sequence and have seeds that don't overlap (since the alternative would be seed=0 for worker 0, seed=1 for worker 1 etc. but seeding like this will cause some overlap when you seed with seed=1 for worker 0).
See `seed_generator` for more info.
In practice you should:
- seed worker 0, get the resulting seed
- seed worker 1 with the resulting seed, get the resulting seed
- etc

Reply:
we can use the iterative approach